### PR TITLE
[WB-5739] no longer run docker as root

### PR DESCRIPTION
**Pull request by stephchen**
https://wandb.atlassian.net/browse/WB-5739

Description
-----------
chown the copied in dir to the current user so we don't need to run as root anymore. also cleans up some args, comments etc around docker stuff

Adds a docker key to the launch_config with a 'user_id' property that allows users to specify a uid that they want to use. If this isn't given, defaults to 1000

Testing
-------

tested on launch

### CODE REVIEW

File: wandb/sdk/launch/docker.py [hunk 1]
```diff
@@ -64,31 +64,30 @@ def generate_docker_image(project: _project_spec.Project, entry_cmd):
...
-        repository_uri=repository_uri, work_dir=project.dir
+        name=name, work_dir=project.dir
     )
 
     wandb_project = project.docker_env["WANDB_PROJECT"]
     wandb_entity = project.docker_env["WANDB_ENTITY"]
     dockerfile = (
         "FROM {imagename}\n"
-        "COPY {build_context_path}/ {workdir}\n"
+        "COPY --chown={uid} {build_context_path}/ {workdir}\n"
```
Comment (stephchen) [inline wandb/sdk/launch/docker.py:84]: ok so we currently copy in the repo twice (once here and once via r2d). should resolve this

File: wandb/sdk/launch/docker.py [hunk 2]
```diff
@@ -64,31 +64,30 @@ def generate_docker_image(project: _project_spec.Project, entry_cmd):
     if not image_id:
         image_id = re.findall(r"Reusing existing image \((.+)\)", stderr)
     if not image_id:
-        raise Exception("error running repo2docker")
+        raise Exception("Error generating docker image")
```
Comment (KyleGoyette) [inline wandb/sdk/launch/docker.py:67]: This is a better error.

File: wandb/sdk/launch/docker.py (hunk 1 above)
Comment (KyleGoyette) [reply wandb/sdk/launch/docker.py:84]: Good catch. We should check that repo2docker is putting the code in the right place for the docker argument and potentially drop this one.

This also brings up a related point, when a user specifies a docker image, I think we should assume that image contains the code?

File: wandb/sdk/launch/_project_spec.py [hunk 3]
```diff
@@ -33,8 +33,9 @@ def __init__(
     ):
 
         self.uri = uri
-        self.name = name  # todo: what to do for default names
+        self.name = name
         if self.name is None and utils._is_wandb_uri(uri):
+            # default name is {project}_{runid}
```
Comment (KyleGoyette) [inline wandb/sdk/launch/_project_spec.py:38]: since we're naming the run after the old run_id, maybe we should add a launch prefix?

**Review by KyleGoyette — APPROVED** [review]
One small suggestion re:experiment names, but looks good

File: wandb/sdk/launch/docker.py [hunk 4]
```diff
@@ -97,6 +96,7 @@ def build_docker_image(project: _project_spec.Project, repository_uri, base_imag
         api_key=api.api_key,
         wandb_project=wandb_project,
         wandb_entity=wandb_entity,
+        uid=os.getuid(),
```
Comment (vanpelt) [inline wandb/sdk/launch/docker.py:99]: We may want to make this configurable.  I believe the jupyter base images always create a user named jovyan with UID 1000.  They then add that user to the root group, we should likely just piggy back on that and let the user change the uid if they choose to.

File: wandb/sdk/launch/docker.py [hunk 5]
```diff
@@ -125,22 +125,69 @@ def build_docker_image(project: _project_spec.Project, repository_uri, base_imag
...
+                env_vars[user_entry] = system_var
+
+    if volumes is not None:
+        for v in volumes:
+            cmd += ["-v", v]
+
+    for key, value in env_vars.items():
+        cmd += ["-e", "{key}={value}".format(key=key, value=value)]
+    cmd += [image.tags[0]]
+    return cmd
```
Comment (vanpelt) [inline wandb/sdk/launch/docker.py:172]: This would a good place to use shlex_quote

`from six.moves import shlex_quote`

**Review by vanpelt — CHANGES_REQUESTED** [review]

File: wandb/sdk/launch/docker.py (hunk 5 above)
Comment (KyleGoyette) [reply wandb/sdk/launch/docker.py:178]: Added this once the command is converted to a string

File: wandb/sdk/launch/docker.py (hunk 4 above)
Comment (KyleGoyette) [reply wandb/sdk/launch/docker.py]: Wasn't entirely sure what the right way to piggy back here is. So I made the UID configurable but default to using the users uid. Would be happy to clarify if you have something else in mind.

File: wandb/cli/cli.py [hunk 6]
```diff
@@ -958,6 +958,12 @@ def _user_args_to_dict(arguments, argument_type="P"):
     "W&B downloads artifacts from distributed URIs passed to parameters of "
     "type 'path' to subdirectories of storage_dir.",
 )
+@click.option(
+    "--uid",
+    "-u",
+    default=None,
+    help="uid to use for the user, defaults to 1000, since blah blah blah",
```
Comment (adrnswanberg) [inline wandb/cli/cli.py:965]: lol, can we just omit the blah blah blah?

File: wandb/sdk/launch/docker.py [hunk 7]
```diff
@@ -56,44 +57,43 @@ def generate_docker_image(project: _project_spec.Project, entry_cmd):
     stderr = subprocess.run(
         cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE
     ).stderr.decode("utf-8")
+    print(stderr)
```
Comment (adrnswanberg) [inline wandb/sdk/launch/docker.py:60]: This'll only print stuff after the process is done running. Would be cool to tee live output.

Comment (adrnswanberg) [reply wandb/sdk/launch/docker.py:60]: But not necessary for this PR.

File: wandb/sdk/launch/docker.py (hunk 1 above)
Comment (adrnswanberg) [reply wandb/sdk/launch/docker.py:89]: according to usage below, it looks like uid might be None. Should we handle that here?

**Review by adrnswanberg — APPROVED** [review]
lgtm apart from some minor comments.

File: wandb/cli/cli.py [hunk 8]
```diff
@@ -954,6 +954,9 @@ def _user_args_to_dict(arguments, argument_type="P"):
     "W&B downloads artifacts from distributed URIs passed to parameters of "
     "type 'path' to subdirectories of storage_dir.",
 )
+@click.option(
+    "--uid", "-u", default=None, help="uid to use for the user, defaults to users uid",
+)
```
Comment (stephchen) [inline wandb/cli/cli.py:959]: tbh idk how i feel about having this in the args it feels too low level/specific to local. maybe something to specify at the config file level? the same way we'd specify eg anyscale details

File: wandb/sdk/launch/docker.py [hunk 9]
```diff
@@ -49,6 +50,7 @@ def generate_docker_image(project: _project_spec.Project, entry_cmd):
     cmd: Sequence[str] = [
         "jupyter-repo2docker",
         "--no-run",
+        "--user-id={}".format(project.uid),
```
Comment (stephchen) [inline wandb/sdk/launch/docker.py:53]: were you saying this still causes permissions issues? feel like it shouldn't with the r2d copy but we'd also have to sync with this uid on our copy until the double copy thing is fixed

Comment (KyleGoyette) [reply wandb/sdk/launch/docker.py:53]: This line solves the permission issues created when a user sets the uid. Currently we chown with the same uid on the copy stuff, so they are synced.

File: wandb/cli/cli.py (hunk 8 above)
Comment (KyleGoyette) [reply wandb/cli/cli.py:959]: all right removing, and we'll have a the uid be specifyable in the config file under the docker key

**Pull request merged**